src/docs/testing/best-practices.md

```diff
@@ -0,0 +1,174 @@
+when designing your app:
+
+### Call setState thoughtfully
+
+* Avoid repetitive and costly work in `build()` methods since `build()`
+  can be invoked frequently when ancestor Widgets rebuild.
+* Localize the `setState()` call to the part of the subtree whose UI
+  actually needs to change. Avoid calling calling setState() high up in
+  the tree if the change is contained to a small part of the tree.
+* Re-use the same child widget instance when building stops that
+  child subtree from re-building when the parent rebuilds.
+  [PENDING: rephrase]
+
+Also see:
+
+* [Performance considerations](https://master-docs-flutter-io.firebaseapp.com/flutter/widgets/StatefulWidget-class.html#performance-considerations),
+  part of the
+  [StatefulWidget](https://master-docs-flutter-io.firebaseapp.com/flutter/widgets/StatefulWidget-class.html)
+  API doc
+* [Improve your Flutter app performance: split your widgets
+  (with MVP)](https://medium.com/flutter-community/improve-your-flutter-app-performance-split-your-widgets-935f97e93f7d)
+  by Simon Lightfoot
+
+### Apply effects only when needed
+
+Use effects carefully, as they can be expensive. Some of them invoke
+`saveLayer()` behind the scenes, which can be an expensive operation.
+
+{{site.alert.note}}
+**Why is savelayer expensive?**<br>
+Calling `saveLayer()` allocates an offscreen buffer. Drawing content
+into the offscreen buffer may trigger render target switches that
+are particularly slow in older GPUs.
+{{site.alert.end}}
+
+Some general rules when applying specific effects:
+
+* Use the [**Opacity**]({{site.api}}/flutter/widgets/Opacity-class.html)
+  widget only when necessary, but if you must use it, add a
+  RepaintBoundary right below Opacity in the widget tree to prevent
+  unnecessary calls to `saveLayer()`.  See [Transparent
+  image]({{site.api}}/flutter/widgets/Opacity-class.html#transparent-image)
+  in the Opacity API page for an example of applying opacity directly
+  to an image, which is faster than using the Opacity widget.
+* **Clipping** doesn’t call `saveLayer()` (unless explicitly requested with
+  `Clip.antiAliasWithSaveLayer()`) so these operations aren’t as expensive
+  as Opacity, but clipping is still costly, so use with caution. By default,
+  clipping is disabled (`Clip.none`), so you must explicitly enable it when
+  needed. 
+
+Other widgets that may trigger `saveLayer()` and are potentially costly:
+
+* [ShaderMask]({{site.api}}/flutter/widgets/ShaderMask-class.html)
+* [ColorFilter]({{site.api}}/flutter/dart-ui/ColorFilter-class.html)
+* [Chip]({{site.api}}/flutter/material/Chip-class.html)&mdash;may
+  cause call to `saveLayer()` if `disabledColorAlpha != 0xff`
+* [Text]({{site.api}}/flutter/widgets/Text-class.html)&mdash;may
+  cause call to `saveLayer()` if there's an `overflowShader`
+
+Ways to avoid calls to `saveLayer()`:
+
+* To implement fading in an image, consider using the FadeInImage widget,
```

Comment:
surround `FadeInImage` with back-ticks?

Reply:
We generally avoid code fonting class names, though there are exceptions. @kwalrath

Reply:
We've been drifting toward the [Google guidelines](https://developers.google.com/style/code-in-text), which (sadly) say to put class names in code font.

In new writing, that means we code font `Future` (code font), but we can still talk about futures (normal font).